feat: add External Secrets Operator (ESO) integration page (#12978)
pulumi/docs: content/docs/esc/integrations/external-secrets-operator.md

---
title: External Secrets Operator (ESO)
title_tag: Integrate with External Secrets Operator (ESO) | Pulumi ESC
h1: External Secrets Operator (ESO)
meta_desc: Pulumi ESC integrates with the External Secrets Operator (ESO) to manage and deliver secrets in Kubernetes clusters.
weight: 2
menu:
  esc:
    identifier: esc-external-secrets-operator
    parent: esc-integrations
aliases:
- /docs/esc/other-integrations/external-secrets-operator/
---

## Overview

[External Secrets Operator](https://external-secrets.io/latest/) is a Kubernetes operator that integrates external secret management systems with Kubernetes. By using External Secrets Operator, you have several advantages over Kubernetes native secrets:

> - Sensitive data is stored and managed by an external service outside the Kubernetes cluster and this leads to better security and compliance posture.

Reviewer: ```suggestion
- Sensitive data is stored and managed by an external service outside the Kubernetes cluster and this leads to better security and compliance.
```

> - External Secrets Operator supports multiple secret providers, so you can use the same operator to manage secrets from different sources.

Reviewer: ```suggestion
- External Secrets Operator supports multiple different sources, so you can use the same operator to manage secrets and configuration from different sources.
```

> - You can take advantage of the advanced features of the secret provider, such as encryption of data at rest and scenarios like secret rotation.

Reviewer: ```suggestion
- Take advantage of the advanced features of the secret provider, such as encryption of data at rest and scenarios like secret rotation.
```

```bash
helm repo add external-secrets https://charts.external-secrets.io
helm repo update

helm upgrade --install external-secrets external-secrets/external-secrets \
    --namespace external-secrets \
    --create-namespace \
    --wait
```

#### Create secret containing Pulumi access token

```bash
kubectl create secret generic pulumi-access-token -from-literal=PULUMI_ACCESS_TOKEN=${PULUMI_ACCESS_TOKEN} \
    --namespace external-secrets
```

#### Create ClusterSecretStore

Now you can create a `ClusterSecretStore` resource that will tell External Secrets Operator to use Pulumi ESC as a secret provider.

If you want to limit the access by namespace, you can create a `SecretStore` resource instead, which is scoped to a single namespace.

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: external-secrets.io/v1beta1
kind: ClusterSecretStore
metadata:
  name: secret-store
spec:
  provider:
    pulumi:
      organization: ${PULUMI_ORG}
      project: ${PULUMI_PROJECT}
      environment: ${PULUMI_ENV}
      accessToken:
        secretRef:
          name: pulumi-access-token
          key: PULUMI_ACCESS_TOKEN
          namespace: external-secrets
  EOF
```

> Please replace `${PULUMI_ORG}`, `${PULUMI_PROJECT}`, `${PULUMI_ENV}` with your Pulumi organization, project, and environment names.

Reviewer: Is this Pulumi Project or ESC project? If ESC  project, I'll be explicitly in calling out "ESC Project" and "ESC Environment" 

Question: why do we have to specific project like this and not have Environment as a fully qualified name? like "<project-name>/<environment-name>" like we have with out CLI?

Author: Done, changed the names.

From a UX point, it is better to have them seperate. If there is any changes you can easily spot them. Especially if you run some GitOps magic on your cluster.

For demo purposes, we assume that we already have an ESC environment `my-org/my-project/my-env` with a secret `my-secret` that we want to manage using External Secrets Operator.

```yaml
values:

>   hello: world

Reviewer: the example YAML isn't consistent with "my-secret" you called out above? should this also be fn::secret::

```

#### Create ExternalSecret

> Now you can create an `ExternalSecret` resource that will tell External Secrets Operator to fetch the secret from Pulumi ESC. There is also an `ClusterExternalSecret` resource, which is a cluster scoped resource that can be used to manage ExternalSecret resources in specific namespaces.

Reviewer: whenever we use K8 specific language like 'ClusterExternalSecret' can we link them to external documentation?

                project: "dirien",
                environment: "hello-world",
                accessToken: {
                    secretRef: {
                        name: patSecret.metadata.name,
                        key: "PULUMI_ACCESS_TOKEN",
                        namespace: patSecret.metadata.namespace,
                    },
                },
            },
        },
    },
}, { dependsOn: externalSecretsRelease });

const externalSecret = new kubernetes.apiextensions.CustomResource("external-secret", {
    apiVersion: "external-secrets.io/v1beta1",
    kind: "ExternalSecret",
    metadata: {
        name: "secret",
        namespace: "default",
    },
    spec: {
        data: [
            {
                secretKey: "esc-secret",
                remoteRef: {
                    key: "hello",
                }
            }
        ],
        refreshInterval: "20s",
        secretStoreRef: {
            kind: clusterSecretStore.kind,
            name: clusterSecretStore.metadata.name,
        }
    },
}, { dependsOn: externalSecretsRelease });
```

{{% /choosable %}}

{{% choosable language python %}}

```python

> import pulumi

Reviewer 2: Mind running this code through `black`? We usually left-justify Python code to make it easier to read: 

![image](https://github.com/user-attachments/assets/276deacc-18cb-464e-9ea0-d43ebc57d16d)

			ApiVersion: pulumi.String("external-secrets.io/v1beta1"),
			Kind:       pulumi.String("ExternalSecret"),
			Metadata: &metav1.ObjectMetaArgs{
				Name:      pulumi.String("secret"),
				Namespace: pulumi.String("default"),
			},
			OtherFields: kubernetes.UntypedArgs{
				"spec": pulumi.Map{
					"data": pulumi.Array{
						pulumi.Map{
							"secretKey": pulumi.String("esc-secret"),
							"remoteRef": pulumi.Map{
								"key": pulumi.String("hello"),
							},
						},
					},
					"refreshInterval": pulumi.String("20s"),
					"secretStoreRef": pulumi.Map{
						"kind": clusterSecretStore.Kind,
						"name": clusterSecretStore.Metadata.Name(),
					},
				},
			},
		}, pulumi.DependsOn([]pulumi.Resource{externalSecretsRelease}))
		if err != nil {
			return err
		}

		return nil
	})
}
```

{{% /choosable %}}

You can then deploy the Pulumi program using below command as you would normally do with any Pulumi program:

```bash
pulumi up
```

#### Cleanup

To remove the External Secrets Operator and the created resources, you can use the following commands:

```bash
pulumi destroy
```

> ## Advanced Configuration

Reviewer: ```suggestion
## Push Secrets to Pulumi ESC
```
